pure-csi-plugin/templates/node.yaml

```
@@ -0,0 +1,233 @@
+{{ if and (eq .Capabilities.KubeVersion.Major "1") (lt .Capabilities.KubeVersion.Minor "14") }}
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: csidrivers.csi.storage.k8s.io
```

So there is only one of these per cluster correct? What happens if another CSI driver is already installed will this fail?

---

Good question... It is very possible a customer would want more than 1 CSI driver in their cluster to provide different tiers of storage

---

Well then the right thing is to remove this from the chart and assume (correctly) that the cluster has been setup to handle CSI.
So customers trying our chart on 1.13 will have to create this CRD as mentioned in the documentation: https://kubernetes-csi.github.io/docs/csi-driver-object.html

---

Although that page does say you have to have this object, but that you can list drivers, plural, so that implies there can be multiple that use the same object... Maybe we should test this before making a decision on what to do here.

---

We cannot assume that the CRD will be present. As far as a customer is concerned, the first CR should bring in the CRD (or we have to write an operator, but that is overkill).
At most we can try to create the CRD, and if it exists, treat it as not-an-error and proceed.

I advise against asking the customer to create the CRD ...

---

Not sure I understand your point. 
We are testing this in our CI/CD - this chart does work for one driver and fail for multiple drivers.

---

I didnt find a easy way in helm to ignore errors.

---

Hi, One last question .... How is this working ?
a. Chart install installs CRD.
b. Chart install installs CR for Pure's CSI Driver object.

I thought (b) could fail due to a timing issue, that the CRD's resource has not been completely initialized in the api server.

---

On 1.13 
1. Chart install - installs CRD no CR is installed yet. CR is created by the sidecar container cluster-driver-registrar after the installation is complete.

---

We can have a separate CRD.yaml we provide and have documentation that if the CRD is not installed they should run this before running the helm install.

Spending too much time to make this seemless is not worth it in my opinion. I am guessing that this will be part of k8s in later releases anyways.

---

We could just add this to the README: I agree that this is getting to be a time sink on k8s 1.13
This issue does not exist on 1.14 and higher versions since it is a builtin once CSI became beta.

For K8s 1.13 ensure that CSIDriver is enabled by
1. Add  "--feature-gates=CSIDriverRegistry=true" to kubectl and the control-plane. 
2. Create the CSIDriver CRD by running the following:
 $> kubectl create -f https://raw.githubusercontent.com/kubernetes/csi-api/master/pkg/crd/manifests/csidriver.yaml

---

I agree

---

I was able to write a job to run and test if there is no-CRD only then create this. For this release I am only using the README. Will include the job in the helm-chart as a pre-install hook in a follow-on PR